keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,631 @@
+    }
+
+    // IP addresses allocated to the pod with associated metadata.
+    PodIPs []PodIPInfo `json:"podIPs,omitempty" protobuf:"bytes,6,opt,name=podIPs"`
+```
+This internal representation should eventually become part of a versioned API (after a period of deprecation for the singular "PodIP" field).
+
+#### Maintaining Compatible Interworking between Old and New Clients
+Any Kubernetes API change needs to consider consistent interworking between a possible mix of clients that are running old vs. new versions of the API. In this particular case, however, there is only ever one writer of the PodStatus object, and it is the API server itself. Therefore, the API server does not have an absolute requirement to implement any safeguards and/or fixups between the singular PodIP and the plural PodIPs fields as described in the guidelines for pluralizing singular API fields that is included in the [Kubernetes API change quidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md).
+
+However, as a defensive coding measure and for future-proofing, the following API version translation logic will be implemented for the PodIP/PodIPs fields:
+
+##### V1 to Core (Internal) Conversion
+- If only V1 PodIP is provided:
+  - Copy V1 PodIP to core PodIPs[0]
+- Else if only V1 PodIPs[] is provided: # Undetermined as to whether this can actually happen (@thockin)
+  - Copy V1 PodIPs[] to core PodIPs[]
+- Else if both V1 PodIP and V1 PodIPs[] are provided:
+  - Verify that V1 PodIP matches V1 PodIPs[0]
+  - Copy V1 PodIPs[] to core PodIPs[]
+- Delete any duplicates in core PodIPs[]
+
+##### Core (Internal) to V1 Conversion
+  - Copy core PodIPs[0] to V1 PodIP
+  - Copy core PodIPs[] to V1 PodIPs[]
+
+### Awareness of Multiple NodeCIDRs per Node
+As with PodIP, corresponding changes will need to be made to NodeCIDR. These changes are essentially the same as the aformentioned PodIP changes which create the pularalization of NodeCIDRs to a slice rather than a singular and making those changes across the internal representation and v1 with associated conversations.
+
+
+#### kubelet Startup Configuration for Dual-Stack Pod CIDRs
+The existing "--pod-cidr" option for the [kubelet startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) will be modified to support multiple IP CIDRs in a comma-separated list (rather than a single IP string), i.e.:
+```
+  --pod-cidr  ipNetSlice   [IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+Only the first address of each IP family will be used; all others will be logged and ignored.
+
+#### kube-proxy Startup Configuration for Dual-Stack Pod CIDRs
+The existing "cluster-cidr" option for the [kube-proxy startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) will be modified to support multiple cluster CIDRs in a comma-separated list (rather than a single IP string), i.e:
```

Comment:
As above, I think we should move this to phase 2

Reply:
kube-proxy and kubelet prober will work with the default IP as-is at the end of phase 1. Updated implementation plan to reflect that changes to both kube-proxy and kubelet prober will be actioned in phase 2.